pkg/virt-handler/node-labeller/kvm-caps-info-plugin.go

```diff
@@ -0,0 +1,263 @@
+const int CapHypervTime = KVM_CAP_HYPERV_TIME;
+const int CapHypervVpIndex = KVM_CAP_HYPERV_VP_INDEX;
+const int CapHypervTlbflush = KVM_CAP_HYPERV_TLBFLUSH;
+const int CapHypervSendIPI = KVM_CAP_HYPERV_SEND_IPI;
+const int CapHypervSynic = KVM_CAP_HYPERV_SYNIC;
+const int CapHypervSynic2 = KVM_CAP_HYPERV_SYNIC2;
+int msr_list_index(void *data, int n, __u32 *index) {
+    struct kvm_msr_list *msrs = (struct kvm_msr_list*)data;
+    if (n >= msrs->nmsrs) {
+        return 1;
+    }
+    *index = msrs->indices[n];
+    return 0;
+}
+*/
+import "C"
+
+import (
+	"os"
+	"syscall"
+	"unsafe"
+
+	"kubevirt.io/client-go/log"
+)
+
+const (
+	HV_X64_MSR_CRASH_CTL               = 0x40000105
+	HV_X64_MSR_RESET                   = 0x40000003
+	HV_X64_MSR_VP_INDEX                = 0x40000002
+	HV_X64_MSR_VP_RUNTIME              = 0x40000010
+	HV_X64_MSR_SCONTROL                = 0x40000080
+	HV_X64_MSR_STIMER0_CONFIG          = 0x400000B0
+	HV_X64_MSR_TSC_FREQUENCY           = 0x40000022
+	HV_X64_MSR_REENLIGHTENMENT_CONTROL = 0x40000106
+	HV_X64_MSR_VP_ASSIST_PAGE          = 0x40000073
+)
+
+type tmsr struct {
+	nmsrs   C.__u32
+	indices []C.__u32
+}
+
+type CapScan struct {
+	supportedMSRS map[uint32]bool
+	fd            uintptr
+	logger        *log.FilteredLogger
+}
+
+type capability struct {
+	Name      string
+	Extension uintptr
+	MSR       int
+}
+
+func NewCapScanner() *CapScan {
+	capScan := new(CapScan)
+	capScan.logger = log.DefaultLogger()
+	return capScan
+}
+
+var CapsDesc = []capability{
+	{
+		Extension: uintptr(C.CapHyperv),
+		Name:      "base",
+	},
+	{
+		Extension: uintptr(C.CapHypervTime),
+		Name:      "time",
+	},
+	{
+		Extension: uintptr(C.CapHypervVpIndex),
+		MSR:       HV_X64_MSR_VP_INDEX,
+		Name:      "vpindex",
+	},
+	{
+		Extension: uintptr(C.CapHypervTlbflush),
+		Name:      "tlbflush",
+	},
+	{
+		Extension: uintptr(C.CapHypervSendIPI),
+		Name:      "ipi",
+	},
+	{
+		Extension: uintptr(C.CapHypervSynic),
+		MSR:       HV_X64_MSR_SCONTROL,
+		Name:      "synic",
+	},
+	{
+		Extension: uintptr(C.CapHypervSynic2),
+		MSR:       HV_X64_MSR_SCONTROL,
+		Name:      "synic2",
+	},
+	{
+		MSR:  HV_X64_MSR_TSC_FREQUENCY,
+		Name: "frequencies",
+	},
+	{
+		MSR:  HV_X64_MSR_RESET,
+		Name: "reset",
+	},
+	{
+		MSR:  HV_X64_MSR_VP_RUNTIME,
+		Name: "runtime",
+	},
+	{
+		MSR:  HV_X64_MSR_STIMER0_CONFIG,
+		Name: "synictimer",
+	},
+	{
+		MSR:  HV_X64_MSR_REENLIGHTENMENT_CONTROL,
+		Name: "reenlightenment",
+	},
+}
+
+func availableMsrs(fd uintptr) ([]uint32, error) {
+
+	// Find our list of MSR indices.
+	// A page should be more than enough here,
+	// eventually if it's not we'll end up with
+	// a failed system call for some reason other
+	// than E2BIG (which just says n is wrong).
+	msrIndices := make([]byte, 4096, 4096)
+	msrs := make([]uint32, 0, 0)
+	for {
+
+		_, err := kvmIOCtl(
+			fd,
+			uintptr(C.IoctlGetMsrIndexList),
+			uintptr(unsafe.Pointer(&msrIndices[0])))
+		if err == syscall.E2BIG {
```

Comment:
This looks like a hack. You could use 2 calls to ioctl:
1. With `kvm_msr_list.nmsrs` equal 0 and no array allocated. The ioctl call would return the size of array needed.
2. Then allocate the struct `kvm_msr_list` with enough size to fit the returned data.

Please, also remove the loop and use 2 calls to `kvmIOCtl()` instead. It would be more visible that the call is only executed twice.